### PR TITLE
Ibc

### PR DESCRIPTION
**Pull request by baidang201**

### CODE REVIEW

File: src/developer-documentation/ibc/ibc.md
Comment (uchu) [inline src/developer-documentation/ibc/ibc.md:488]: Is there a reason to have `// ///` here for commenting?

File: src/developer-documentation/ibc/ibc.md
Comment (uchu) [inline src/developer-documentation/ibc/ibc.md:560]: It there a reason to keep commented blocks for documentation?

File: src/developer-documentation/ibc/ibc.md
Comment (uchu) [inline src/developer-documentation/ibc/ibc.md:163]: Is that something not finished? Do we need this in docs?

File: src/developer-documentation/ibc/ibc.md
Comment (uchu) [inline src/developer-documentation/ibc/ibc.md:2]: Can we use capital letters where possible for general terms like IBC, NFT, etc?

File: src/developer-documentation/ibc/ibc.md
Comment (uchu) [inline src/developer-documentation/ibc/ibc.md:168]: TODO looks fine in code, not sure about docs.

**Review by uchu — COMMENTED** [review]
Just minor comments and questions

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:168]: ok, i will remove it

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:2]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:163]: ics20 interface and PSP37 interface  is finish
 
 ics721/ics27   is in design
 
 and we can remove  no relationship  struct

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:560]: no need

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:488]: no, i will remove later

**Review by uchu — APPROVED** [review]
LGTM. Although I'd make some markdown cleanup after it's merged

**Comment by uchu** [comment]
@baidang201 I am ok to merge it after another pending PR is merged, as it has some structural changes.

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:22]:
```suggestion
```rust
/// Transfer token; this allows us to transfer *exactly one* native token.
pub fn execute_transfer(
    &self,
    msg: TransferMsg,
    amount: Amount,
    sender: Addr,
) -> Result<Response, Error>;

/// Return the port ID bound by this contract.
pub fn query_port(&self) -> PortResponse;
```
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:25]:
```suggestion
```rust
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:99]:
```suggestion
```rust
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:98]:
```suggestion
## 2.2 Struct
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:183]:
```suggestion
```rust
pub struct Addr(String);

pub struct Coin {
    pub denom: String,
    pub amount: u128,
}

pub struct Cw20Coin {
    pub address: String,
    pub amount: u128,
}

pub struct TransferMsg {
    /// The local channel to send the packets on
    pub channel: String,
    /// The remote address to send to.
    /// Don't use HumanAddress as this will likely have a different Bech32 prefix than we use
    /// and cannot be validated locally
    pub remote_address: String,
    /// How long the packet lives in seconds. If not specified, use default_timeout
    pub timeout: Option<u64>,
    /// An optional memo to add to the IBC transfer
    pub memo: Option<String>,
}

pub enum Amount {
    Native(Coin),
    Cw20(Cw20Coin),
}

impl Amount {
    pub fn from_parts(denom: String, amount: u128) -> Self {
        if denom.starts_with("cw20:") {
            let address = denom.get(5..).unwrap().into();
            Amount::Cw20(Cw20Coin { address, amount })
        } else {
            Amount::Native(Coin { denom, amount })
        }
    }

    pub fn cw20(amount: u128, addr: &str) -> Self {
        Amount::Cw20(Cw20Coin {
            address: addr.into(),
            amount: amount,
        })
    }

    pub fn native(amount: u128, denom: &str) -> Self {
        Amount::Native(Coin {
            denom: denom.to_string(),
            amount: amount,
        })
    }
}

impl Amount {
    pub fn denom(&self) -> String {
        match self {
            Amount::Native(c) => c.denom.clone(),
            Amount::Cw20(c) => "cw20:".to_owned() + c.address.as_str(),
        }
    }

    pub fn amount(&self) -> u128 {
        match self {
            Amount::Native(c) => c.amount,
            Amount::Cw20(c) => c.amount,
        }
    }

    /// convert the amount into u64
    pub fn u64_amount(&self) -> Result<u64, Error> {
        Ok(self.amount().try_into().unwrap())
    }

    pub fn is_empty(&self) -> bool {
        match self {
            Amount::Native(c) => c.amount == 0,
            Amount::Cw20(c) => c.amount == 0,
        }
    }
}
```
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:225]:
```suggestion
```rust
pub enum Error {
    /// StdError
    StdError,
    /// PaymentError
    PaymentError,
    /// AdminError
    AdminError,
    /// #[error("Channel doesn't exist: {id}")]
    NoSuchChannel { id: String },
    /// #[error("Didn't send any funds")]
    NoFunds {},
    /// #[error("Amount larger than 2**64, not supported by ics20 packets")]
    AmountOverflow {},
    /// #[error("Only supports channel with ibc version ics20-1, got {version}")]
    InvalidIbcVersion { version: String },
    /// #[error("Only supports unordered channel")]
    OnlyOrderedChannel {},
    /// #[error("Insufficient funds to redeem voucher on channel")]
    InsufficientFunds {},
    /// #[error("Only accepts tokens that originate on this chain, not native tokens of remote chain")]
    NoForeignTokens {},
    /// #[error("Parsed port from denom ({port}) doesn't match packet")]
    FromOtherPort { port: String },
    /// #[error("Parsed channel from denom ({channel}) doesn't match packet")]
    FromOtherChannel { channel: String },
    /// #[error("Cannot migrate from different contract type: {previous_contract}")]
    CannotMigrate { previous_contract: String },
    /// #[error("Cannot migrate from unsupported version: {previous_version}")]
    CannotMigrateVersion { previous_version: String },
    /// #[error("Got a submessage reply with unknown id: {id}")]
    UnknownReplyId { id: u64 },
    /// #[error("You cannot lower the gas limit for a contract on the allow list")]
    CannotLowerGas,
    /// #[error("Only the governance contract can do this")]
    Unauthorized,
    /// #[error("You can only send cw20 tokens that have been explicitly allowed by governance")]
    NotOnAllowList,
}
```

```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:227]:
```suggestion
[IBC Interfaces for CosmWasm Contracts](https://github.com/CosmWasm/cosmwasm/blob/main/IBC.md)
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:229]:
```suggestion
[Contracts Semantics](https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md)
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:231]:
```suggestion
[Introduction to CosmWasm](https://github.com/CosmWasm/cosmwasm/blob/main/README.md)
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:233]:
```suggestion
[Channel and Packet Semantics](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics)
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:235]:
```suggestion
[PSP37 Protocol](https://github.com/w3f/PSPs/blob/master/PSPs/psp-37.md)
```

File: src/developer-documentation/ibc/ibc.md
Comment (xcthulhu) [inline src/developer-documentation/ibc/ibc.md:237]:
```suggestion
[PSP37 Trait](https://github.com/Brushfam/openbrush-contracts/blob/main/contracts/src/traits/psp37/psp37.rs)
```

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:22]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:25]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:98]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:183]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:225]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:229]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:231]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:233]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:235]: sure

File: src/developer-documentation/ibc/ibc.md
Comment (baidang201) [reply src/developer-documentation/ibc/ibc.md:237]: sure

**Comment by uchu** [comment]
@xcthulhu @baidang201 Are all questions addressed and we a good to merge it?

**Pull request merged**
